### PR TITLE
[rdbms] Schema migration tool

### PR DESCRIPTION
**Pull request by marcoguerri**
Schema migration tool which supports both pure SQL migration
files and go migrations (for more complex operations which
require back-filling and manipulating previous entries).
Based on github.com/pressly/goose.
    
Every migration is executed within a transaction. SQL migrations
offer the possibility to turn off transactions, while golang migrations
by default receive a *sql.Tx object. At the moment, this cannot be
changed.

### CODE REVIEW

**Comment by marcoguerri** [comment]
Follow-up proposal after the discussion in https://github.com/facebookincubator/contest/pull/148.

More details on the changes included in this PR:

* Plugin registration has been moved away from main, so that it can be easily re-used in the main package of the migration tool
* ` db/rdbms/migration` now contains two migration which add support for #118 . In particular, see `0002_migrate_descriptor_to_extended_descriptor.go` as an example of what a go migration would look like
* In order to enable migration for #118 , I had to backport some structure definition from #118 to this PR. Those forward declarations are contained in several `forward_decl.go`. Once #118 is merged, they should be cleaned up.

In terms of test plan:
1) I checked out #118 (on top of which I added some minor changes, first and foremost the new field in the db is now a dedicated `extended_descriptor` column). Running `status <JOB_ID>` results in the following error:
```
"Error": "failed to fetch request for job ID 1: could not fetch job request: could not get job request with id 1: Error 1054: Unknown column 'extended_descripto
r' in 'field list'
```

2) Applied migration 0001, which adds the `extended_descriptor` field. Result of `status <JOB_ID` is now:
```
 "Error": "failed to fetch request for job ID 1: could not fetch job request: could not get job request with job id 1: sql: Scan error on column index 5, name \"
extended_descriptor\": converting NULL to string is unsupported"
```
3) Applied migration 0002, which backfills previous job entries. `status <JOB_ID` succeeds with relevant output.

**Comment by marcoguerri** [comment]
I am having a not so great time with `go.mod`. Please for the moment ignore integ tests failures, while I try to sort them out.

**Comment by marcoguerri** [comment]
Builds is now fine, test coverage is problematic. I'd love to get some feedback on this approach before going down the route of adding additional tests.

File: cmds/plugins/plugins.go
Comment (tfg13) [inline cmds/plugins/plugins.go:59]:
```suggestion
	// sample function to prove that function registration works.
```

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (tfg13) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:107]: I was recently told one needs to check `rows.Err()`, to tell apart legit ends of  `rows.Next()` from error cases (https://golang.org/pkg/database/sql/#Rows)

**Review by tfg13 — COMMENTED** [review]
LGTM, but I will leave the final stamp to @xaionaro and @dimkup, who have been very active in the original PR and probably have more context. I only found some minor things. Ping me if you need a stamp to unblock.

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:245]: **nitpicking:**

It could be `%w`.

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:64]: **nitpicking:**

1. Extra `float64()`
2. It could've been `d.Seconds() / 1000` or `float64(d.Nanoseconds()) / float64(time.Millisecond.Nanoseconds())`. Personally I would prefer `d.Seconds() / 1000`.

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:49]: **nitpicking:**

A typo: `s/Schmema/Schema/`

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:44]: **nitpicking:**

A typo: `s/submmitted/submitted/`

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:199]: **nitpicking:**

I would call it `stepDescFetched` instead of `stepDescFetcher`. Otherwise it was confusing me for few seconds while reading the code :)

File: cmds/plugins/plugins.go
Comment (xaionaro) [inline cmds/plugins/plugins.go:65]: **warning:**

I haven't found any usage for this function.

File: cmds/plugins/plugins.go
Comment (xaionaro) [inline cmds/plugins/plugins.go:66]: Could you provide an example where these `userFunctions` are required? Why not to use a function directly? Is it for templates?

File: db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go
Comment (xaionaro) [inline db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go:249]: **nitpicking:**

May be it makes sense to add `rows.Err()` to the error message.

File: go.sum
Comment (xaionaro) [inline go.sum:241]: **nitpicking:**

And obsolete checksum (for version 2.6.0). May be it makes sense to execute `go mod tidy`.

File: tools/migration/rdbms/main.go
Comment (xaionaro) [inline tools/migration/rdbms/main.go:88]: **nitpicking:**

Since it is only one field, it could've been `logger := log.WithField("migration", filepath.Base(m.Name))`

**Review by xaionaro — APPROVED** [review]
LGTM. There's only one concern: the so-called `userFunctions` which I do not understand why are required.

**Comment by marcoguerri** [comment]
`userFunctions` is a copy-paste from previous `main.go`. IIRC, they  were helper functions usable in templates. I'll review if they are actually used anywhere, they might be dead code.

Meanwhile, I pushed an update which adds tests for the migration, logic has not changed.

**Comment by tfg13** [comment]
Oh yeah we use `userFunctions` a ton for templates, please keep them :P

**Comment by marcoguerri** [comment]
Ok, after some iterations, this is finally the corrected coveraged:

```
integration | 60.86% <51.40%> (-1.20%) 
unittests | 25.73% <0.00%> (-0.01%)
```

Now, `codecov/patch` is still complaining, despite the yaml changes. Trying to figure out why.

**Comment by marcoguerri** [comment]
Applied suggestions from @xaionaro. Finally ready to go.

**Pull request merged**
